src/pages/TestingPage/TestControls/useControlForm.ts

```diff
@@ -0,0 +1,156 @@
+import { useEffect, useReducer, useState } from "react";
+
+type InputData = {
+    id: string;
+    type: string;
+    value: number | null;
+    enabled: boolean;
+    validity: { isValid: boolean; msg: string };
+};
+
+type ChangingValue = {
+    id: string;
+    value: number | null;
+};
+
+type EnablingValue = {
+    id: string;
+    enabled: boolean;
+};
+
+export type FormData = Array<InputData>;
+
+export type Form = { formData: FormData; isValid: boolean };
+
+type SubmitHandler = () => void;
+type ChangeValue = (id: string, value: number) => void;
+type ChangeEnable = (id: string, enable: boolean) => void;
+
+type Action =
+    | { type: "CHANGE VALUE"; payload: ChangingValue }
+    | { type: "CHANGE ENABLE"; payload: EnablingValue }
+    | { type: "RESET INITIAL STATE"; payload: Form };
+
+const searchId = (form: Form, id: string): number => {
+    let index = form.formData.findIndex((inputData) => inputData.id == id);
+    return index;
+};
+
+const checkType = (type: string, value: number): boolean => {
+    //TODO: which types?
+    return true;
+};
+
+const checkIsValid = (formData: FormData): boolean => {
+    const result =
+        formData.reduce(
+            (prev, currentInput) =>
+                prev &&
+                (currentInput.validity.isValid || !currentInput.enabled),
+            true
+        ) &&
+        formData.some((input) => input.enabled) &&
+        !formData.some((input) => input.enabled && !input.value);
+    return result;
+};
```

Comment:
the two some function are a bit redundant. This function can be simplified.

Reply:
they are different conditions, I think they are best understood separately